src/trs/rewrite.rs

```diff
@@ -386,72 +338,22 @@ impl TRS {
-    /// # extern crate programinduction;
-    /// # extern crate rand;
-    /// # extern crate term_rewriting;
-    /// # use programinduction::trs::{TRS, Lexicon};
-    /// # use rand::{thread_rng};
-    /// # use term_rewriting::{Context, RuleContext, Signature, parse_rule, parse_term};
-    /// # fn main() {
-    /// let mut sig = Signature::default();
-    ///
-    /// let t = parse_term(&mut sig, "A(B(x_ x_))").expect("parse of A(B(x_ x_))");
-    /// let r = parse_rule(&mut sig, "C(y_) = B(y_ y_)").expect("parse of C(y_) = B(y_ y_)");
-    /// let mut rng = thread_rng();
-    ///
-    /// let result = TRS::inverse_evaluation_helper(&r, &t, &mut rng);
-    ///
-    /// assert_eq!(result.display(), "A(C(x_))");
-    /// # }
-    /// ```
-    pub fn inverse_evaluation_helper<R: Rng>(rule: &Rule, t: &Term, rng: &mut R) -> Term {
+    fn inverse_evaluation_helper<R: Rng>(rule: &Rule, t: &Term, rng: &mut R) -> Term {
         let rhs_idx = rng.gen_range(0, rule.rhs.len());
         TRS::replace_term_helper(t, &rule.rhs[rhs_idx], rule.lhs.clone())
     }
     /// Given two rules, attempts to apply one rule inversely to the other.
-    ///
-    /// # Example
-    ///
-    /// ```
-    /// # #[macro_use] extern crate polytype;
-    /// # extern crate programinduction;
-    /// # extern crate rand;
-    /// # extern crate term_rewriting;
-    /// # use programinduction::trs::{TRS, Lexicon};
-    /// # use rand::{thread_rng};
-    /// # use term_rewriting::{Context, RuleContext, Signature, parse_rule, parse_term};
-    /// # fn main() {
-    /// let mut sig = Signature::default();
-    ///
-    /// let r = parse_rule(&mut sig, "A(B(x_ x_)) = D B(x_ x_)").expect("parse of A(B(x_ x_)) = D B(x_ x_)");
-    /// let rule = parse_rule(&mut sig, "C(y_) = B(y_ y_)").expect("parse of C(y_) = B(y_ y_)");
-    /// let mut rng = thread_rng();
-    ///
-    /// let result = TRS::inverse_evaluate_rule_helper(&rule, &r, &mut rng);
-    ///
-    /// assert_eq!(result.unwrap().pretty(), "A(C(x_)) = D C(x_)");
-    /// # }
-    /// ```
-    pub fn inverse_evaluate_rule_helper<R: Rng>(
-        rule: &Rule,
-        r: &Rule,
-        rng: &mut R,
-    ) -> Option<Rule> {
+    fn inverse_evaluate_rule_helper<R: Rng>(rule: &Rule, r: &Rule, rng: &mut R) -> Option<Rule> {
         let lhs = TRS::inverse_evaluation_helper(rule, &r.lhs, rng);
         let mut rhs: Vec<Term> = vec![];
         for idx in 0..r.rhs.len() {
             rhs.push(TRS::inverse_evaluation_helper(rule, &r.rhs[idx], rng));
         }
         Rule::new(lhs, rhs)
     }
-    /// Selects two Rules from the TRS at random and atempts to inverse evaluate one rule on the other, if it
-    /// succeeds it takes that new rule and inserts it imediately after the background.
-    pub fn inverse_evaluate<R: Rng>(&self, rng: &mut R) -> Result<TRS, SampleError> {
+
+    /// Extracts a sngle rule from the TRS selecting only 1 rhs
+    fn sample_rule<R: Rng>(&self, rng: &mut R) -> Result<Rule, SampleError> {
```

Comment:
Funny - I forgot that I'd put this in the last review and just wrote a similar function, `TRS::choose_clause`, a couple days ago. I suppose we have a strong use case for it! I'm not sure why this one is quite so long, though. Here's my implementation:

```rust
    fn choose_clause<R: Rng>(&self, rng: &mut R) -> Result<Rule, SampleError> {
        let rules = {
            let num_rules = self.len();
            let background = &self.lex.0.read().expect("poisoned lexicon").background;
            let num_background = background.len();
            &self.utrs.rules[0..(num_rules - num_background)]
        };
        let mut clauses = rules.iter().flat_map(Rule::clauses).collect_vec();
        let idx = (0..clauses.len())
            .choose(rng)
            .ok_or(SampleError::OptionsExhausted)?;
        Ok(clauses.swap_remove(idx))
    }
```

Am I missing something?

Reply:
Oh, I was reading some of the lines you'd deleted as part of the current function. They're pretty similar in length.